### PR TITLE
ConcurrentQueue 128-byte cache line

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

**Comment by sdmaclea** [comment]
@dotnet/arm64-contrib @stephentoub @kouvel PTAL

**Comment by stephentoub** [comment]
Performance tests/results to justify the change?

**Comment by sdmaclea** [comment]
No only implicit evidence that these hot cache lines impacted performance for 64-byte cache lines and were therefore worthy of optimizing.  There is no reason to believe the same is not true for 128-byte line caches.

I will measure performance impact of the equivalent change in CoreCLR on plaintext.  Since the CoreCLR code is a copy of this code, I wanted to keep them in sync.

**Comment by stephentoub** [comment]
I understand, but this change isn't free; it's increasing the size of each segment by 192 bytes.  That may not be a big deal, but it's at least worth verifying is worthwhile.  How common are 128-byte cache lines?

File: src/System.Threading.Tasks.Dataflow/src/Internal/ProducerConsumerQueues.cs [hunk 1]
```diff
@@ -551,7 +551,7 @@ static class PaddingHelpers
     }
 
     /// <summary>Padding structure used to minimize false sharing in SingleProducerSingleConsumerQueue{T}.</summary>
-    [StructLayout(LayoutKind.Explicit, Size = PaddingHelpers.CACHE_LINE_SIZE - sizeof(Int32))] // Based on common case of 64-byte cache lines
```
Comment (stephentoub) [inline src/System.Threading.Tasks.Dataflow/src/Internal/ProducerConsumerQueues.cs]: Why did you delete this comment?

Comment (sdmaclea) [inline src/System.Threading.Tasks.Dataflow/src/Internal/ProducerConsumerQueues.cs]: I was out of date and inconsistent with the 128 byte value of CACHE_LINE_SIZE.

Comment (stephentoub) [inline src/System.Threading.Tasks.Dataflow/src/Internal/ProducerConsumerQueues.cs]: Ah, I didn't see that the existing value was already set to 128.

**Comment by sdmaclea** [comment]
> How common are 128-byte cache lines?

I am not sure.  Most Qualcomm mobile arm32 & arm64 use 128-byte cache lines.  All Qualcomm Datacenter parts currently use 128-byte cache lines, I believe PowerPC also used.  

In CoreCLR mscorlib there is `src/mscorlib/shared/System/Span.NonGeneric.cs:#if CORECLR && (AMD64 || ARM64)`, so there is precedence for making optimizations platform dependent.  Not sure if this would be acceptable in CoreFX.

**Comment by stephentoub** [comment]
> Many Qualcomm mobile arm32 & arm64 use 128-byte cache lines.  All Qualcomm Datacenter parts currently use 128-byte cache lines, I believe PowerPC also used.

Ok, thanks.

> In CoreCLR mscorlib there is src/mscorlib/shared/System/Span.NonGeneric.cs:#if CORECLR && (AMD64 \|\| ARM64), so there is precedence for making optimizations platform dependent.  Not sure if this would be acceptable in CoreFX.

There's a single build of System.Collections.Concurrent.dll, regardless of OS, architecture, etc.  We'd need to change that, which is quite expensive from a build and deployment perspective.  And even if we did, to my knowledge we don't have any way of saying "use this particular build on machines with this cache line size"... we'd need to assume a particular size based on a particular architecture.  Doesn't seem worthwhile.

I'd love to see some data that validates this change is worthwhile on 128-byte cache lines.  It very likely is, but any time we add cost, we should validate it's for a good reason.  Assuming it is, I'm fine with the change.

**Comment by SunnyWar** [comment]
Every time I see changes like this I think of the promise of .Net when it was introduced to optimize for the platform the program is running on. Given this, it seem odd to hard-code the cache line size. Can it be done at runtime, or better yet, by the jitter? [https://stackoverflow.com/questions/794632/programmatically-get-the-cache-line-size](url)

**Comment by kouvel** [comment]
Would it make sense to move all of the concurrent collections to CoreLib? All of them are used by CoreLib and it seems redundant to have a copy there and here. Since CoreLib is already built for different architectures, it would make it easy to special-case this for the architecture.

**Comment by stephentoub** [comment]
> it would make it easy to special-case this for the architecture.

I'm not clear how special-casing for the arch is valuable here, as the cache line size isn't tied to an architecture.  Am I missing something?

**Comment by stephentoub** [comment]
> All of them are used by CoreLib 

ConcurrentBag isn't used, is it?

ConcurrentStack is used only by PinnableBufferCache, and honestly I'm not sure that's the right choice for it, given CS's allocation profile.  Separately we should revisit that.

ConcurrentDictionary and ConcurrentQueue are both used, though only a relatively small portion of their surface area.

That said, I don't have a strong opinion about where the types live; we could certainly look at moving them if it was worthwhile.

**Comment by tannergooding** [comment]
> Since CoreLib is already built for different architectures, it would make it easy to special-case this for the architecture.

I would, personally, rather see a general way to say: Automatic Layout with Padding based on Cache Line size.

Users outside of CoreFX/CoreCLR should be able to implement their own "optimized" data structures where the can force any alignment and do "intelligent" things, like padding structures based on the cache line size/etc.

**Comment by stephentoub** [comment]
> I would, personally, rather see a general way to say: Automatic Layout with Padding based on Cache Line size.

At one point we investigated adding a `[CacheLineAligned]` attribute that could be used on fields that the runtime would ensure were always aligned on a cache-line boundary, but it never got productized.  I would also like a solution along those lines.

**Comment by sdmaclea** [comment]
@stephentoub ARM64 Ubuntu perf data: For plaintext, the CoreCLR equivalent of this change results in a 17% improvement on my tip.

**Comment by tannergooding** [comment]
> At one point we investigated adding a [CacheLineAligned] attribute

@stephentoub, is there an existing issue for this? If not, I'm going to log one 😄

**Comment by stephentoub** [comment]
> is there an existing issue for this?

I don't believe so.

**Comment by tannergooding** [comment]
@stephentoub: https://github.com/dotnet/corefx/issues/22790

**Comment by stephentoub** [comment]
> the CoreCLR equivalent of this change results in a 17% improvement on my tip.

Ok, thanks.  Let's go with this for now, and hopefully the extra memory consumption will be inconsequential.

**Comment by stephentoub** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build please

**Comment by hanblee** [comment]
> I understand, but this change isn't free; it's increasing the size of each segment by 192 bytes, even on (the vast majority?) of architectures with 64-byte cache lines.

> ARM64 Ubuntu perf data: For plaintext, the CoreCLR equivalent of this change results in a 17% improvement on my tip.

> Ok, thanks. Let's go with this for now, and hopefully the extra memory consumption will be inconsequential.

Since this is a performance related PR affecting common code, it would be good to understand performance impact on other platforms.

@sdmaclea Do you have any data that show the change does not regress performance on x64?

**Comment by stephentoub** [comment]
> Do you have any data that show the change does not regress performance on x64?

It will regress, at least in terms of memory consumption, with every segment being larger than it used to be.  It's a good point that in theory it could regress throughput as well (due to needing to load more memory), though I doubt it would in any meaningful way.

**Comment by hanblee** [comment]
> It will regress, at least in terms of memory consumption, with every segment being larger than it used to be. It's a good point that in theory it could regress throughput as well (due to needing to load more memory), though I doubt it would in any meaningful way.

Understood. What I am saying is that for performance related PRs, they need to be data driven - sometimes what we expect to see is not what we get.

**Comment by sdmaclea** [comment]
> performance on x64?

@hanblee No measurable impact expected.  No data collected.

**Comment by karelz** [comment]
What is status of this PR? No update for last 2+ weeks.

I would love to see perf tests checked in along side this change, so that we can independently verify the improvement and no impact on other platforms. Is that reasonable ask?

**Comment by sdmaclea** [comment]
@karelz I will work on adding a perf test next week.  I will create a new file `src/System.Collections.Concurrent/tests/ConcurrentQueuePerfTest.netcoreapp.cs`

**Comment by sdmaclea** [comment]
Added a `ConcurrentQueue` perf test.  Code is derived from `Concurrent_Clear_NoExceptions` in the functional tests.  Code compiles.  I still have to figure out the incantation to run it and see perf measurements.

**Comment by sdmaclea** [comment]
Also note I did not attempt to force false sharing in the ConcurrentQueue segments (by forcing a specific bad alignment).  So perf change on arm64 might not be repeatable.

**Comment by karelz** [comment]
@sdmaclea what's next step? I see no activity for last week ...

**Comment by sdmaclea** [comment]
@karelz 
> so that we can independently verify the improvement and no impact on other platforms

I assumed someone wanted to test other platforms.  I have verified this is beneficial using the CoreCLR version of this patch.  If you want me to I can try to figure out how to run the uBenchmark.

**Comment by sdmaclea** [comment]
@karelz I was also expecting a code review

**Comment by karelz** [comment]
> I assumed someone wanted to test other platforms. I have verified this is beneficial using the CoreCLR version of this patch. If you want me to I can try to figure out how to run the uBenchmark.

In general, we expect every submitter of a PR (MS, member, non-member) to see it through end-to-end, on all affected platforms, including verification that there is no regression. Functional is covered by CI legs. Perf and other concerns (e.g. stress) are manual by submitter (in a form which can be potentially verified by others if desired).
If there was different arrangement I am not aware of, please let me know ...

**Comment by SunnyWar** [comment]
@karelz > In general, we expect every submitter of a PR (MS, member, non-member) to see it through end-to-end, on all affected platforms, 

LOL. you must be joking.

**Comment by karelz** [comment]
> LOL. you must be joking.

Not really. Why do you think so?

We always held all developers responsible for their code changes end-to-end and from all aspects (that holds for open-source and closed-source .NET source code, even prior to any OSS efforts on our side).
When there are concerns during code review, we expect the author to address it. If the author doesn't have the right expertise/HW, it is of course ok to seek help/advice/team-up with someone/etc., but the driver is still the author.

Such end-to-end means that the author has to make sure all tests pass. If there are unrelated test failures (proved to be unrelated noise), we expect the author to seek help/file bugs (sometimes other engineers jump in when they notice, but it is not a service we guarantee).
If there is perf concern during code review, or reliability concern, or compat concern, or test gap concern, or end-to-end scenario impact concern, then it is up to the author to provide enough evidence/test assets to address those concerns. Sometimes more people chime in, but in 99.9% we try to get to consensus between author and all reviewers.
Even if the change is the "right fix", but it uncovers other bugs up the stream which depend on the bad behavior, the author of the "right fix" is responsible to make sure it is all addressed (which can be done with help of others of course). We don't allow checking in code when we know it breaks something in the product, even if it is "the right change for the component". The end-to-end product is what matters most.

It is a bar we hold for all our Microsoft engineers and I don't see any reason why we should soften it significantly for anyone else. Of course, we try to help more external contributors, or new team members, provide guidance and make sure everyone is successful, but we do not have a magic "test team" which would go and make sure someone's changes are ok prior to checkin -- code reviews are supposed to call out concerns and it is up to author to address them.
Of course, we do monitor quality of the product (i.e. post-checkin), so if something gets through unnoticed (even code reviews are not perfect and can miss some aspects/interaction between components), we do address such things (which might mean reverting the change which caused the troubles).

It all feels very natural to me, so I am curious if you have different view point.

**Comment by sdmaclea** [comment]
@karelz I agree in principle to most of what you are saying.  The comment came off as harsh and off topic.  

You had requested a micro-benchmark in an earlier comment because you said it would allow others to independently verify the performance impact of this change.  When you asked for next steps, I assumed this verification was a step you intended.  I was simply saying the test was ready to be used/tried.  

So given your comment, the the next step is a code review.   I will also in parallel try to run the new perf test on arm64.

**Comment by sdmaclea** [comment]
@dotnet-bot help

**Comment by sdmaclea** [comment]
@karelz I see there are no perf tests in CI.  Looks like this is issue #19200.  If you want to include in dev expectations, then #19200 is critical.

**Comment by kouvel** [comment]
Sorry for the delay, I'll take a look shortly

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    public class Perf_ConcurrentQueueTests
+    {
+        /// <summary>
+        /// Creates a list containing a number of elements equal to the specified size
+        /// </summary>
+        [Benchmark]
+        [InlineData(8, 1000)]
+        [InlineData(50, 1000)]
+        public void Enqueue_Dequeue(int threadsCount, int itemsPerThread)
+        {
```
Comment (kouvel) [inline src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: Most of the [Benchmark] tests are looping over iterations similarly to the example here: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/performance-tests.md

Probably need to do that here along with calling iteration.StartMeasurement().

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            var q = new ConcurrentQueue<int>();
+            Task.WaitAll((from i in Enumerable.Range(0, threadsCount) select Task.Run(() =>
```
Comment (kouvel) [inline src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: The number of tasks that actually run concurrently will be limited by thread pool heuristics, I suggest using dedicated threads instead or using long-running tasks that don't run on the thread pool:
```c#
Task.Factory.StartNew(() =>
{
   ...
}, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
```

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                var random = new Random();
```
Comment (kouvel) [inline src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: May be good to use a specific seed to keep the test deterministic

File: src/System.Collections.Concurrent/tests/Performance/System.Collections.Concurrent.Performance.Tests.csproj [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  <PropertyGroup>
+    <IncludePerformanceTests>true</IncludePerformanceTests>
+    <ProjectGuid>{13A6510E-E11C-4329-BCDD-4DE561AB6CC0}</ProjectGuid>
+    <DisableTests Condition="'$(TargetGroup)' == 'uap' AND ('$(ArchGroup)' == 'arm' OR '$(ArchGroup)' == 'arm64')">true</DisableTests>
+  </PropertyGroup>
+  <!-- Default configurations to help VS understand the configurations -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' " />
+  <ItemGroup Condition="'$(DisableTests)' != 'true'">
+    <Compile Include="ConcurrentQueuePerfTest.netcoreapp.cs" />
```
Comment (kouvel) [inline src/System.Collections.Concurrent/tests/Performance/System.Collections.Concurrent.Performance.Tests.csproj]: I don't think the .netcoreapp suffix is necessary, the test itself doesn't require it, it's just not enabled on others for reasons external to this file.

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs (hunk 3 above)
Comment (sdmaclea) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: @kouvel Do you have an example (I don't write much C# core) ?  I copied this code from other multi-threaded corefx perf test.

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs (hunk 4 above)
Comment (sdmaclea) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: OK, but It will be hard to keep the test deterministic with many threads contending and sharing the same random number generator.

Comment (kouvel) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: Yea true, probably won't make much difference

**Comment by kouvel** [comment]
On x64, I'm seeing no change in perf on your test. On my ConcurrentQueueThroughput test in https://github.com/dotnet/coreclr/pull/13670, I'm seeing an improvement:
```
                                                   Left score       Right score      ∆ Score %
-------------------------------------------------  ---------------  ---------------  ---------
ConcurrentQueueThroughput 1Pc                      37769.40 ±1.31%  39750.25 ±0.63%      5.24%
-------------------------------------------------  ---------------  ---------------  ---------
```

**Comment by stephentoub** [comment]
@kouvel, thanks for checking.  Presumably your machine has 64-byte cache lines?  I'd think the best we could hope for is that this doesn't regress perf in that case (so I'm a little surprised there was a small bump in one of the tests... maybe just noise?)

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs (hunk 3 above)
Comment (kouvel) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: Could follow the general pattern in https://github.com/dotnet/corefx/blob/master/src/System.Collections/tests/Performance/Perf.List.cs. Each iteration could be everything inside the Task.WaitAll.

**Comment by karelz** [comment]
> The comment came off as harsh and off topic.

Not sure if you mean any of my previous answers. If yes, I apologize. Text is not the best way to provide all context, doubts and meanings :(.

> You had requested a micro-benchmark in an earlier comment because you said it would allow others to independently verify the performance impact of this change. When you asked for next steps, I assumed this verification was a step you intended.

I see. Sorry for the confusion - my original ask (more a suggestion) for micro-benchmark was targeted as "having evidence" and giving anyone chance to review the micro-benchmark itself (incl. its completeness / coverage) and also giving anyone the option to replicate same results if that person chooses to do so.
Hopefully it makes sense now. If not, let me know.

> I will also in parallel try to run the new perf test on arm64.

If the code could affect perf also on x86 and amd64 (in a different way than arm32 and arm64), we will need verification on those platforms as well prior to checkin. I will let final decision on the reviewers of the change.

> I see there are no perf tests in CI. Looks like this is issue #19200. If you want to include in dev expectations, then #19200 is critical.

This is more complicated than one would like to. Let me try to explain:
First, #19200 is about Desktop (full .NET Framework) in closed-source perf lab. Having that implemented would not help.
Second, we do not have perf test results currently published anywhere publicly. It is likely on perf team long-term roadmap (at least I had it there 2+ years ago when I was perf lead). The truth is there has been TONS of higher-priority things since than in perf area and it was seldom a blocker for devs (they can run perf tests locally in the worst case).
Third, you raise a good point that it would be awesome to have perf tests integrated into CI. Again, it is on backlog, with current workaround to run it locally for ad-hoc new perf tests (like yours) and offline monitoring of perf lab results internally in closed-source (for unintended horrible regressions which might slip through).
Fourth, monitoring of perf tests in general is very costly thing. Creating a perf tests which is stable is hard. And even such stable perf tests can be influenced by infrastructure and monitoring them is non-trivial and costly (and we're talking about extremely controlled environment) -- which leads to reducing the number of perf tests a team wants to run and monitor on regular basis. As such it is rare that "just add more perf tests into regular perf runs" is the right answer cost/benefit-wise. Another implication is that coverage of scenarios in regular perf runs is limited and therefore we rely more on ad-hoc targeted perf benchmarks for PRs which are perf sensitive ... I know this is complicated topic and I'll be happy to chat offline even more. Sadly there is no right answer and I believe that current situation (asking PR authors to verify changes via targeted perf tests when there is perf-impact concern) is reasonable prioritization.

**Comment by kouvel** [comment]
@stephentoub, I tried on two different processors, both with 64-byte cache line size and I'm consistently seeing a 5-8% improvement on both.

**Comment by stephentoub** [comment]
> I'm consistently seeing a 5-8% improvement on both.

Any theories as to why?

**Comment by kouvel** [comment]
I can't think of a reason, maybe the extra separation is allowing it to somehow/accidentally parallelize the memory accesses better

**Comment by stephentoub** [comment]
> I can't think of a reason

Ok, thanks.  Given that this improves things for a known reason on some systems, and at least doesn't regress throughput elsewhere (it'll regress memory overheads, but it shouldn't be significant compared to the arrays being used), so I'm fine with it.

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs (hunk 3 above)
Comment (sdmaclea) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: I do not see any thread usage in the example you cite.  I'll use https://github.com/dotnet/coreclr/pull/13670 as a sample for thread usage.

Comment (sdmaclea) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: `SemaphoreSlimWaitDrainRate()` looks like a good sample

**Comment by sdmaclea** [comment]
@dotnet-bot help

File: src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs (hunk 3 above)
Comment (kouvel) [reply src/System.Collections.Concurrent/tests/Performance/ConcurrentQueuePerfTest.netcoreapp.cs]: Regarding Perf.List.cs, I just meant for that to be an example for a [Benchmark] test.

Many of the tests I wrote in https://github.com/dotnet/coreclr/pull/13670 have a high error margin and they're not all necessarily bad to regress if there's compensation elsewhere so I didn't make them official [Benchmark] tests that are verified by the CI. There's also a ConcurrentQueueThroughput test there.

**Comment by sdmaclea** [comment]
I did revise the draft perf test per @kouvel suggestion.  However I removed it based on:
+ @kouvel 's ConcurrentQueueThroughput test in dotnet/coreclr#13670
+ The test is untested.  `./build-managed.sh -release -tests -- /p:Performance=true` returns and error on this and every other Performance.Test in the tree.  This is roughly the command used in perf.groovy
+ "@"dotnet-bot "help" has no indication on how to run perf tests.

I left it in the PR history in case someone wants to revive

**Comment by kouvel** [comment]
LGTM, thanks!
